src/helics/core/mpi/MpiComms.cpp

```diff
@@ -9,38 +9,319 @@ Lawrence Livermore National Laboratory, operated by Lawrence Livermore National
+
+bool MpiComms::mpiCommsExists = false;
+std::mutex MpiComms::mpiSerialMutex;
+int MpiComms::commRank = -1;
+
+MpiComms::MpiComms ()
 {
+    if (mpiCommsExists)
+    {
+        std::cout << "WARNING: MPIComms object already created, unexpected results may occur" << std::endl;
+    }
+    std::cout << "MpiComms()" << std::endl;
+    initMPI();
+    mpiCommsExists = true;
+}
+
+MpiComms::MpiComms(const int &brokerRank)
+    : brokerRank(brokerRank)
+{
+    if (mpiCommsExists)
+    {
+        std::cout << "WARNING: MPIComms object already created, unexpected results may occur" << std::endl;
+    }
+    std::cout << "MpiComms(" << std::to_string(brokerRank) << ")" << std::endl;
+    initMPI();
+    if (brokerRank == -1)
+    {
+        setBrokerRank (commRank);
+    }
+    mpiCommsExists = true;
 }
+
 /** destructor*/
-MpiComms::~MpiComms () = default;
+MpiComms::~MpiComms ()
+{
+    std::cout << "Disconnecting MPIComms" << std::endl;
+    disconnect ();
+    std::cout << "Finished disconnect" << std::endl;
 
-void MpiComms::queue_rx_function () {}
+    int mpi_initialized;
+    MPI_Initialized(&mpi_initialized);
+    if (mpi_initialized)
+    {
+        std::cout << "Finalizing MPI" << std::endl;
+        std::lock_guard<std::mutex> lock(mpiSerialMutex);
+        MPI_Finalize();
+        shutdown = true;
+        std::cout << "MPI Finalized" << std::endl;
+    }
+}
 
-void MpiComms::closeTransmitter ()
+bool MpiComms::initMPI()
 {
-    ActionMessage rt (CMD_PROTOCOL);
-    rt.index = DISCONNECT;
-    transmit (-1, rt);
+    std::lock_guard<std::mutex> lock(mpiSerialMutex);
+
+    // Initialize MPI with MPI_THREAD_SERIALIZED
+    int mpi_initialized;
+    int mpi_thread_level;
+
+    MPI_Initialized(&mpi_initialized);
+
+    if (!mpi_initialized)
+    {
+        MPI_Init_thread(nullptr, nullptr, MPI_THREAD_SERIALIZED, &mpi_thread_level);
+
+        MPI_Initialized(&mpi_initialized);
+        if (!mpi_initialized)
+        {
+            std::cerr << "MPI initialization failed" << std::endl;
+            return false;
+        }
+    }
+
+    MPI_Query_thread(&mpi_thread_level);
+
+    if (mpi_thread_level < MPI_THREAD_SERIALIZED)
+    {
+        std::cerr << "MPI_THREAD_SERIALIZED support required" << std::endl;
+        return false;
+    }
+
+    return true;
 }
 
-void MpiComms::closeReceiver ()
+void MpiComms::serializeSendMPI(std::vector<char> message, int dest, int tag, MPI_Comm comm) {
+    std::cout << "SendMPI from " << getAddress() << " to " << dest << std::endl;
+    std::lock_guard<std::mutex> lock(mpiSerialMutex);
+    MPI_Request req;
+    MPI_Isend(message.data(), message.size(), MPI_CHAR, dest, tag, comm, &req);
+
+    int message_sent = false;
+    while (!message_sent)
+    {
+        MPI_Test(&req, &message_sent, MPI_STATUS_IGNORE);
+    }
+}
+
+std::vector<char> MpiComms::serializeReceiveMPI(int src, int tag, MPI_Comm comm) {
+    int message_waiting = false;
+    MPI_Status status;
+
+    while (!message_waiting && !shutdown)
```

Comment:
is this a busy loop or is there some sort of waiting going on?

Reply:
It is a busy loop -- do you think a call to sleep should be added in? The Iprobe call used in receive is a non-blocking check to see if any messages are waiting to be received, so if there is one Irecv can be called -- the MPI_Test call then checks to see if the call to Irecv has finished copying the received data into the buffer. This could also be restructured to get rid of the send/receive methods and instead use a separate thread that uses queues for incoming/outgoing messages and keeps all the MPI calls during normal operation in one place, eliminating the need for the mutex other than when exiting.

Reply:
Actually, with multiple comm objects the mutex would still be needed to serialize the different threads using MPI calls.

Reply:
All the other comm types have some sort of way to wake up the thread when a message is received. Sleep isn't going to help here as that would significantly increase latency.  But the way you described it, if you can't really have multiple calls from the same process going on at the same time, I think we would need to put all the calls into a single thread, and have the busy loop deal with the transmission and reception.  

longer term you said there was another level of MPI that had some more threading capabilities but it wasn't available commonly, if that is the case then we probably want code that switches to that mode if available and falls back if not.

Reply:
As I've been reading more of the MPI documentation and papers on MPI threading, it seems like you're right about switching to a single MPI communication thread that gets shared by all MPIComms objects is the way to go, and to avoid multiple threads entirely. MPI's threaded mode is slower than single/funneled mode (how much depends on the system+MPI implementation) since they have locks and other synchronization happening internally with the threaded mode.

Reply:
I am guess most applications use a well defined pattern of sends and expected received, rather than an asynchronous pattern.  I am guessing this is going to take some experimentation once we get it running, especially once we get to larger scales.